fvtest/compilertriltest/IficmpTest.cpp

```diff
@@ -0,0 +1,174 @@
+ */
+TEST_P(IfcmpReductionTest, Reduction)
+   {
+   auto param = GetParam();
+
+   const char *type = std::get<0>(param);
+   const char *sign = std::get<1>(param);
+   const char *lege = std::get<2>(param);
+   const char *val  = std::get<3>(param);
+
+   char inputTrees[600] = {0};
+   std::snprintf(inputTrees, 600,
+         R"((method return="NoType"
+            (block name="b" fallthrough="u"
+              (%sstore temp="val"                        ; int val = 65535;
+                 (%sconst 65535) )
+                (if%s%scmp%s target="t"
+                    (%sand
+                        (%sload temp="val")
+                        (%sconst 16))
+                    (%sconst %s) ) )
+            (block name="u" (return))
+            (block name="t" (return))
+            ) ; method
+            )"
+       , type, type, type, sign, lege, type, type, type, type, val);
+
+   auto trees = parseString(inputTrees);
+
+   ASSERT_NOTNULL(trees);
+
+   Tril::JitBuilderCompiler compiler{trees};
+
+   ASSERT_EQ(0, compiler.compile()) << "Compilation failed";
+   }
+
+/*
+* Create a name for each test instead of using numbers.
+* E.g., LongUnsignedGeReducible denotes a case, in which:
+* - The type is long unsigned
+* - The operation is greater than or equals (so the opcode is iflucmpge)
+* - The reduction is possible
+*/
+std::string reductionTestNameGenerator(::testing::TestParamInfo<IfcmpReductionParamType> paramInfo)
+    {
+    std::string name;
+    auto param = paramInfo.param;
+
+    if (std::string("i") == std::get<0>(param))
+        {
+        name += "Integer";
+        }
+    else if (std::string("l") == std::get<0>(param))
+        {
+        name += "Long";
+        }
+    else if (std::string("b") == std::get<0>(param))
+        {
+        name += "Byte";
+        }
+    else
+        {
+        name += "Short";
+        }
+
+    if (std::string("u") == std::get<1>(param))
+        {
+        name += "Unsigned";
+        }
+    else
+        {
+        name += "Signed";
+        }
+
+    if (std::string("ge") == std::get<2>(param))
+        {
+        name += "Ge";
+        }
+    else
+        {
+        name += "Le";
+        }
+
+    if (std::string("16") == std::get<3>(param))
+        {
+        name += "Reducible";
+        }
+    else
+        {
+        name += "Nonreducible";
+        }
+    return name;
+    }
+
+/* Tests are combinations of the following:
+ * 1) Integer / long / address / byte / short
+ * 2) Signed / unsigned (e.g., ificmp / ifiUcmp)
+ * 3) Less than or equal / greater than or equal
+ * 4) Reduction can be performed (the const value in "and" is equal to the value in "ifcmp") / reduction cannot be performed
+ */
+INSTANTIATE_TEST_CASE_P(IfcmpReduction, IfcmpReductionTest,
+        ::testing::Combine(
+            ::testing::Values("i", "l", "b", "s"),
+            ::testing::Values("u", ""),
+            ::testing::Values("ge", "le"),
+            ::testing::Values("16", "17")),
```

Comment:
Can you add some more test cases covering what was observed above with `if ((x & 3) <= 3)` for example. We should also cover a variety of corner cases here.